### DISCUSSION TITLE
[v5] Can't retrieve FormData with request.all() with autoprocess false

### DISCUSSION THREAD

**Discussion by D-Marc1** [adonisjs/core#1861] [Help]
I'm using Adonis 5. I can't retrieve `FormData` (multipart/form-data) with `request.all()` if autoProcess is false for streaming uploads. You used to be able to get the body in v4 with `request.multipart.field`, but this doesn't exist in v5. Is this a bug in v5 or not yet implemented?

**Comment by D-Marc1** [comment]
@thetutlage Sorry, don't mean to bother you, but just trying to confirm if this is a bug or not, in case I should raise an issue for this.

**Reply by thetutlage** [reply]
When you say `autoProcess = false`, then simply you are asking AdonisJS to not touch the request body stream and you will manually process it and hence `request.all()` doesn't have any data to provide. 

The following code should work

```ts
Route.post('/', async ({ request }) => {
  request.multipart.onFile('file', {}, async (part) => {
    // skipping the stream for demo
    part.resume()
  })
  await request.multipart.process()

  // Stream has been processed and hence you can access the fields now
  return request.all()
})
``` 

Also, I don't think self handling a stream that has files + regular fields is a great idea. Usually one should use direct streaming for exclusive file uploads ( an endpoint that just upload files and nothing else ).

If you compare this flow with [S3 direct uploads](https://softwareontheroad.com/aws-s3-secure-direct-upload/), the process is pretty much the same. 

- You have a form with a some fields + a file input
- The file is uploading directly to s3 as a separate request
- Once the file upload is finished, you get some kind of file id and then you submit the form along with the file id.
- So form submission + file upload is handled separately. 

Now, if you swap `s3 direct upload` with AdonisJS raw stream handling. The flow should remain the same. A dedicated endpoint piping files (as streams) to a 3rd party like s3 and form submission is handled separately.

**Reply by D-Marc1** [reply]
Thank you so much for the reply and info. But how would I able to get the body before processing the file in v5? in v4, you used to be able to do the following and you were able to retrieve the body before processing the file.

```js
const body = {};

request.multipart.field((name, value) => {
  body[name] = value;
});
```

Is there no way of retrieving the body before the processed file in v5?

I do take your point about the philosophy of having the endpoint solely being for uploading in most cases. But I do think there's certain cases where it's necessary to combine text and files, like when a transaction needs to happen. If this functionality no longer exists in v5, like it did in v4, can I suggest it as a feature request?

**Reply by thetutlage** [reply]
You cannot get the body without processing the stream. This is what `request.multipart.process` does. Files + Fields are part of the same stream.

Also in v4, you were collecting body manually and now it is collected automatically, that is the only difference and rest is all the same.

In the following example you have the body

```ts
await request.multipart.process()
console.log(request.all())
```

**Reply by D-Marc1** [reply]
Could I request having an option to collect the body manually then? Otherwise, there's no way of retrieving the request data before processing the file, like you were able to in v4. It seems like v5 is adding convenience at the expense of removing functionality that existed in v4, as you can only retrieve the request text data after the file is processed. 

For anyone who googles this, right now, my solution is to simply pass in the values as query string parameters, but it would be nice be able to achieve this with FormData being passed to the body.

**Reply by thetutlage** [reply]
I am not sure how you can collect the values before as there is no guarantee that the field will be emitted before the file. Again, it will be nice if you can share a concrete example

**Reply by thetutlage** [reply]
> It seems like v5 is adding convenience at the expense of removing functionality that existed in v4, as you can only retrieve the request text data after the file is processed

This is simply an assumption you are making without understanding the internals. In v4 the following code works by co-incidence

```ts
request.multipart.field((name, value) => {
  body[name] = value;
});

request.multipart.file('name', () => {
  console.log(body)
});
```

You are assuming that the `field` callback will always be executed before the `file` callback and that is not 100% guaranteed. It depends upon how the multipart data is put together by the client.

**Reply by D-Marc1** [reply]
Appreciate the super fast replies! Thanks for letting me know. I actually didn't know this. I should probably check back on projects I've done in v4 then.

An example for why I would need this would be a situation where I'm inserting data into a database first or using text input values for the file names or folders. There's some cases where it's useful to wrap it in a transaction, so the inserted data would get rolled back if the file upload fails. Consider the following:

```js
const requestData = request.all();

await Database.transaction(async (trx) => {
  const user = await User.create(requestData, { client: trx });
  
  const { id } = user.$attributes;

  request.multipart.onFile('file', {}, async (file) => {
    const params = {
      Bucket: 'my-bucket',
      Key: `users/${id}.jpg`,
      Body: file,
    };

    const s3 = new AWS.S3();

    await s3.upload(params).promise();
  });

  await request.multipart.process();
});
```
